[GeoMechanicsApplication] Refactor processing of DOFs by elements and conditions

avdg81 opened the pull request:
**📝 Description**
This PR eliminates several repetitions of the order of degrees of freedom (in elements and conditions). Ideally, the order should be defined just once. This will improve the maintenance of the corresponding code. Also, it will become much easier to change the order as needed.

As a bonus, due to removing the duplications, we have also addressed a significant amount of code smells.

avdg81 committed:
commit-msg.txt

avdg81 committed:
Improved the functionality of `ExtractEquationIdsFrom`

It now returns the correct result when passing a non-empty vector of DOF
pointers.

Also removed a redundant semicolon.

avdg81 committed:
The U-Pw base element extracts the equation IDs using the new function

avdg81 committed:
The U-Pw diff order element extracts the equation IDs from the DOFs

avdg81 committed:
The transient Pw element extracts the equation IDs from the DOFs

avdg81 committed:
The transient T element extracts the equation IDs from the DOFs

avdg81 committed:
The transient Pw interface element extracts the equation IDs from the DOFs

avdg81 committed:
The truss element base extracts the equation IDs from the DOFs

avdg81 committed:
The structural element base extracts the equation IDs from the DOFs

avdg81 committed:
`GetValuesVector` no longer repeats the DOF order

Instead, it now asks for the element's degrees of freedom and it fetches
the corresponding values, except for water pressures, which are set to
zero (why??).

Other changes include:
- `GetDofList` forwards the request to the new (private) member function
  `GetDofs`. The latter does not require a const-ref to a `ProcessInfo`
  instance. For now, the implementation of `GetDofList` has been moved
  as-is to `GetDofs`.
- Simplified the implementation of `EquationIdVector` by using
  `GetDofs` rather than `GetDofList`.
- Unused parameter are now unnamed.

avdg81 committed:
`GetFirstDerivativesVector` no longer repeats the DOF order

Instead, it now asks for the element's degrees of freedom and for each
DOF it gets the corresponding first time derivative, except for first
time derivatives of water pressure, which are always set to zero
(why??).

avdg81 committed:
`GetSecondDerivativesVector` no longer repeats the DOF order

Instead, it now asks for the element's degrees of freedom and for each
DOF it gets the corresponding second time derivative, except for second
time derivatives of water pressure, which are always set to zero
(why??).

Also renamed a local variable.

avdg81 committed:
Moved `ExtractEquationIdsFrom` to `dof_utilities.{cpp,h}`

avdg81 committed:
Reimplemented getting the DOFs of the U-Pw diff order condition

avdg81 committed:
Don't repeat the DOF order when getting the equation IDs

The U-Pw diff order condition now reuses the DOF utility function for
that.

avdg81 committed:
Reimplemented the function that gets the DOFs of Pw condition

avdg81 committed:
Don't repeat the DOF order when getting the equation IDs

The Pw condition now reuses the DOF utility function for that.

avdg81 committed:
Reimplemented getting the DOFs of the temperature condition

avdg81 committed:
Don't repeat the DOF order when getting the equation IDs

The thermal condition now reuses the DOF utility function for that.

avdg81 committed:
Reimplemented the function that gets the DOFs of U-Pw condition

avdg81 committed:
Don't repeat the DOF order when getting the equation IDs

The U-Pw condition now reuses the DOF utility function for that.

avdg81 committed:
Several consistency improvements

- Added `GetDofs` to class `GeoStructuralBaseElement`.
- Simplified `GeoStructuralBaseElement<...>::EquationIdVector`.
- Simplified `PwCondition<...>::GetDofs`.
- Simplified `GeoTCondition<...>::GetDofs`.
- Cleaned up the implementation of `UPwBaseElement<...>::GetDofs`.

avdg81 committed:
A few more consistency improvements

avdg81 committed:
Added a utility function for extracting values from DOFs

Other changes include:
- Moved getting the DOFs of a truss element to a new member function and
  simplified the implementation.
- Simplified the implementation of
  `GeoTrussElementBase<...>::EquationIdVector`.
- Unused function parameters should be unnamed.
- Renamed a function parameter.

avdg81 committed:
The structural base element now gets its DOF values from the helper

avdg81 committed:
Extracted getting the DOFs for transient Pw elements

Also simplified the member function that extracts the equation IDs for
the DOFs of the transient Pw element.

avdg81 committed:
Extracted getting the DOFs for transient Pw interface elements

Also simplified the member function that extracts the equation IDs for
the DOFs of the transient Pw interface element.

avdg81 committed:
Extracted getting the first derivatives of any U-Pw element

Given the degrees of freedom of a U-Pw element (in any order), this new
function returns the corresponding first derivatives. In line with the
previous implementation, this function always returns 0.0 for water
pressure DOFs.

Also added a few comments about zero vectors being returned where actual
values would be expected.

avdg81 committed:
Extracted getting the DOFs of the small strain diff order U-Pw element

Also made sure that some unused function parameters are unnamed.

avdg81 committed:
Extracted getting the DOFs of the transient thermal element

avdg81 committed:
Several code improvements

- Extracted a function that gets the first time derivatives for a
  sequence of degrees of freedom.
- Removed several redundant occurrences of `this->`.
- Swapped two statements for consistency.
- Restored a blank line, which caused an unintended difference.

avdg81 committed:
Reused the function that extracts the first time derivatives

avdg81 committed:
Reused the function that extracts the first time derivatives for U-Pw

Also made a function parameter unnamed, since it was not used. And
simplified the implementation of another function.

avdg81 committed:
Extracted a function that gets the second derivatives of U-Pw elements

avdg81 committed:
Reused the function that gets the second derivatives of U-Pw elements

avdg81 committed:
Added a function for extracting values of U-Pw DOFs

So far this new function is used exclusively by the Lysmer absorbing
boundary condition. To make this work, the `GetDofs` member function of
class `UPwCondition` had to be made `protected`.

Also removed a few redundant occurrences of `this->`.

avdg81 committed:
Reused the function that gets the first time derivatives for U-Pw DOFs

Now also the Lysmer absorbing condition uses this function. The function
has been renamed to make it consistent with the function that extracts
the DOFs's values.

avdg81 committed:
Reused the function that gets the solution step values for U-Pw DOFs

Also renamed the function that extracts the second time derivatives for
U-Pw DOFs (to make it consistent with the other function names).

avdg81 committed:
Renamed a function to make it consistent with other function names

avdg81 committed:
Added a function to extract the second time derivative values

So far, the function is used exclusively by the "geo" structural element
base.

avdg81 committed:
Reused the function to extract the second time derivative values

avdg81 committed:
Several minor improvements

- Restored a blank line that was accidentally removed.
- Removed a few more redundant occurrences of `this->`.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:
```diff
@@ -0,0 +1,116 @@
+#include "dof_utilities.h"
+#include "includes/variables.h"
+
+namespace Kratos
```
To avoid conflicts with other applications, I'd propose to make this a separate namespace
```suggestion
namespace Kratos::DofUtilities
```

avdg81 replied:
Good suggestion. I have made the namespace even more specific: `Kratos::Geo::DofUtilities`.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:
```diff
@@ -0,0 +1,116 @@
+Vector ExtractSolutionStepValuesOfUPwDofs(const std::vector<Dof<double>*>& rDofs, int Step)
+{
+    auto result        = Vector(rDofs.size());
+    auto get_dof_value = [Step](const auto p_dof) -> double {
```
I think this could be const (or does the transform take a non-const ref?)
```suggestion
    const auto get_dof_value = [Step](const auto p_dof) -> double {
```

avdg81 replied:
The `auto` here deduces the type of the lambda expression. So far, I have never encountered a lambda expression that was assigned to a `const auto` instance. I have browsed through the examples in the C++ Core Guidelines, but they all seem to use just `auto`, so without the `const`. I would suggest to follow those examples, unless you can provide me with an authoritative reference that actually prescribes to use `const auto` for these cases.

rfaasse replied:
I was indeed to quick with this comment, I had the same comment for the lambdas in the dof utils, but came to the same conclusion that it does not make too much sense for a lambda expression and removed them again. Totally agree with you that we keep it as is

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.h:
```diff
@@ -78,6 +66,15 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoTCondition : public Condition {
+                       [](const auto& r_node) { return r_node.pGetDof(TEMPERATURE); });
+        return result;
+    }
+
```
In Pw_condition the GetDofs implementation is in the .cpp file ( for me the expected place. ) Here it is in the .h file.
Their implementation is necessarily very similar. I wonder whether it can be a utility implementation where the D.O.F. variable name or a list of variable names is passed ( WATER_PRESSURE, TEMPERATURE etc. )

avdg81 replied:
Yes, I have created a utility function that returns the degrees of freedom for the given collection of nodes and a single DOF variable. I have not implemented the same functionality given a _list_ of DOF variables. The reason is that in the current situation we need to accommodate two orderings for U-Pw elements: (1) where displacements and water pressures are intertwined [for the non-diff order elements], and (2) where all displacement DOFs come first followed by the water pressure DOFs [for the diff order elements]. I'd propose to revisit this after we have made the DOF ordering consistent across all U-Pw elements.

Also, I did move the implementation from the header file to the implementation file, as you suggested.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp:
```diff
@@ -178,71 +179,14 @@ void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::CalculateDampingMatrix(Matrix
-    }
-
-    KRATOS_CATCH("")
+    rValues = ExtractFirstTimeDerivativesOfUPwDofs(this->GetDofs(), Step);
```
Step here is probably 1 for start of the step ( previous value ), 0 for end of the step ( current value ). Apart from the confusing inverse setup of 0 and 1, maybe the name Step can be improved.

avdg81 replied:
I agree with you that a name such as `BufferIndex` would be much more accurate and descriptive. However, we inherit this parameter name from the Kratos core (see `condition.h`). So if we would like to change it, I feel we'd better start from there. I did, however, adopt the parameter name `BufferIndex` while writing the new DOF utility functions.

Also, there is an issue with the virtual function itself, since it has a default value for parameter `Step`, which opens the door to having different default values for different overrides. See for instance [this rule of the C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c140-do-not-provide-different-default-arguments-for-a-virtual-function-and-an-overrider), which warns for that. Clang-tidy also reports it:

![image](https://github.com/KratosMultiphysics/Kratos/assets/28814468/1d40baa5-6772-4280-972d-d4738e68135c)

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp:
```diff
@@ -70,34 +71,9 @@ void GeneralUPwDiffOrderCondition::Initialize(const ProcessInfo& rCurrentProcess
-
-    for (SizeType i=0; i<NumPNodes; ++i)
-        rConditionDofList[Index++] = GetGeometry()[i].pGetDof( WATER_PRESSURE );
-
-    KRATOS_CATCH( "" )
+    rConditionDofList = GetDofs();
```
I wonder whether this should have an implementation or should it be abstract and always overridden by a specific implementation.

avdg81 replied:
As indicated before, we might be able to simplify things even more after we have rearranged the DOFs. But since we intend to replace these element classes (as part of the element redesign), we'd better do this right when we implement the new elements.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp:
```diff
@@ -780,6 +639,19 @@ unsigned int UPwBaseElement<TDim, TNumNodes>::GetNumberOfDOF() const
 }
 
+template <unsigned int TDim, unsigned int TNumNodes>
+Element::DofsVectorType UPwBaseElement<TDim, TNumNodes>::GetDofs() const
```
This will probably be the only remaining implementation for the UPw element when the whole redesign is complete.

avdg81 replied:
Yes, that makes sense.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/geo_structural_base_element.cpp:
```diff
@@ -601,6 +467,23 @@ SizeType GeoStructuralBaseElement<TDim, TNumNodes>::GetAlongNumberIntegrationPoi
+            result.push_back(r_node.pGetDof(ROTATION_X));
+            result.push_back(r_node.pGetDof(ROTATION_Y));
+        }
+        result.push_back(r_node.pGetDof(ROTATION_Z));
```
Which makes clear that the structural base element is a beam element.

avdg81 replied:
Agreed. This is something to keep an eye on when we move over to the structural elements of the `StructuralMechanicsApplication` (since the "geo" structural elements will be removed at some point in time).

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:
```diff
@@ -0,0 +1,116 @@
+    };
+    std::transform(rDofs.begin(), rDofs.end(), result.begin(), get_second_time_derivative);
+    return result;
+}
```
On first glance these are four repetitions of a similar thing. The remark about VariableData I don't understand yet. As far as I have understood all D.O.F. are double by nature.

avdg81 replied:
We have discussed this in-person. The bottom line is that the functions ending with `OfUPwDofs` are only there to not change the current behavior. However, we do need to investigate why the water pressure (and its time derivatives) are assumed to be zero. When we can show that this is not needed at all (or if we can find a better alternative), then we can start changing the function's behavior. And, in fact, that may mean we no longer need these dedicated functions.

Also, I have rephrased the comment about the `VariableData`. Could you please check whether it is clearer now?

WPK4FEM reviewed (commented):
Hi Anne,
That condensed functionality considerably, thank you. Some parts left that I don't properly understand, maybe spoken explanation helps there.
Wijtze Pieter

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/Pw_condition.cpp:
```diff
@@ -153,6 +123,16 @@ void PwCondition<TDim,TNumNodes>::
+{
+    auto       result = Condition::DofsVectorType{};
+    const auto nodes  = GetGeometry();
+    std::transform(nodes.begin(), nodes.end(), std::back_inserter(result),
```
It's nice to see these examples of the stl library

avdg81 replied:
Thank you :-)

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.h:
```diff
@@ -78,6 +66,15 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoTCondition : public Condition {
                               const ProcessInfo& rCurrentProcessInfo);
 
 private:
+    [[nodiscard]] DofsVectorType GetDofs() const
```
Could we move this function to the cpp?

avdg81 replied:
Wijtze Pieter had the exact same remark, and I totally agree with it. Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.h:
```diff
@@ -46,28 +46,16 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) GeoTCondition : public Condition {
 
-    void GetDofList(DofsVectorType& rConditionDofList,
-                    const ProcessInfo& ) const override
+    void GetDofList(DofsVectorType& rConditionDofList, const ProcessInfo&) const override
```
Could we move this function to the cpp?

avdg81 replied:
Good catch. Yes, we can. Done.

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_lysmer_absorbing_condition.cpp:
```diff
@@ -178,71 +179,14 @@ void UPwLysmerAbsorbingCondition<TDim, TNumNodes>::CalculateDampingMatrix(Matrix
-            KRATOS_ERROR << "undefined dimension in GetValuesVector... illegal operation!!" << this->Id() << std::endl;
-        }
-    KRATOS_CATCH("")
+    rValues = ExtractSolutionStepValuesOfUPwDofs(this->GetDofs(), Step);
```
Seeing the inheritance structure (and since `GetValuesVector` is an override from `Condition`), I think it would be better if we move this function to the U_Pw Condition (which is way easier due to your refactoring). However, it is of course a bit out of scope, so I leave it up to you if you think it's good to do it in this PR.

avdg81 replied:
Good suggestion. I would, however, do that in a separate PR, since it might change existing behavior. The aim of this PR is a pure refactoring, so no behavioral changes.

rfaasse replied:
Agreed!

rfaasse commented on applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp:
```diff
@@ -347,4 +297,23 @@ void GeneralUPwDiffOrderCondition::CalculateAndAddConditionForce(VectorType& rRi
+
+    for (const auto& r_node : *mpPressureGeometry) {
+        result.push_back(r_node.pGetDof(WATER_PRESSURE));
+    }
```
I think this loop for the water pressure could be another transform with back_inserter

avdg81 replied:
Done. In fact, I have moved that code (with `std::transform`) to a utility function and call it from here.

rfaasse replied:
Perfect

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:
```diff
@@ -118,6 +93,7 @@ void TransientPwElement<TDim, TNumNodes>::GetValuesVector(Vector& rValues, int S
 
     if (rValues.size() != N_DOF) rValues.resize(N_DOF, false);
 
+    // Why are we constructing a zero vector here?
```
I think we should not keep these comments (but do we have an answer?)

avdg81 replied:
The problem is that at present we don't know why the water pressures (and their time derivatives) are set to zero by these functions. In my opinion, that is beyond the scope of this PR. We should, however, create an issue to investigate it. These new comments serve as a reminder and as a way to quickly find these anomalies when we attempt to address them.

rfaasse replied:
Good point, let's keep them then indeed!

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:
```diff
@@ -669,6 +647,16 @@ unsigned int TransientPwElement<TDim, TNumNodes>::GetNumberOfDOF() const
+    const auto nodes  = this->GetGeometry();
+    std::transform(nodes.begin(), nodes.end(), std::back_inserter(result),
+                   [](const auto& r_node) { return r_node.pGetDof(WATER_PRESSURE); });
+    return result;
```
As @WPK4FEM said, there are indeed quite some versions of getting a 'single' dof (like WATER_PRESSURE/TEMPERATURE), we might extract a function.

avdg81 replied:
Thank you for the suggestion. I have added a utility function (`ExtractDofsFromNodes`) that returns the degrees of freedom of a collection of nodes given a DOF variable.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_interface_element.cpp:
```diff
@@ -729,6 +706,7 @@ void TransientPwInterfaceElement<TDim, TNumNodes>::GetValuesVector(Vector& rValu
 
     if (rValues.size() != N_DOF) rValues.resize(N_DOF, false);
 
+    // Why are we constructing a zero vector here?
```
Also here, it might be better to remove the comments

avdg81 replied:
See my earlier explanation of why I would like to keep them.

rfaasse replied:
Agreed

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.h:
```diff
@@ -300,6 +287,15 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientThermalElement : public Ele
+        DofsVectorType result;
+        for (const auto& r_node : GetGeometry()) {
+            result.push_back(r_node.pGetDof(TEMPERATURE));
+        }
```
Here it can also be an std::transform (or an extracted function)

avdg81 replied:
Yes, I have reused the extracted function (`ExtractDofsFromNodes`) here. I have also searched the other files that I have touched in this PR, and called this new function from there as well.

rfaasse replied:
Thanks, looks very clean!

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:
```diff
@@ -0,0 +1,116 @@
+namespace Kratos
+{
+
+std::vector<std::size_t> ExtractEquationIdsFrom(const std::vector<Dof<double>*>& rDofs)
```
since `EquationId()` is a const function, I think also the Dof<double>* could be const here. Same holds for the other functions if I'm correct.

avdg81 replied:
No, unfortunately, the alias `DofsVectorType` (see, for instance, `element.h`) uses raw pointers to mutable objects:
```
typedef std::vector<DofType::Pointer> DofsVectorType;
```
where `typedef Dof<double> DofType;` and `using Pointer=Dof*;`. So I cannot change this to `std::vector<const Dof<double>*>`. However, as you can see in the parameters of the lambda expressions, I do assign each pointer-to-mutable to a pointer-to-const: `[Step](const auto p_dof) -> double { ... }`. I don't see how I can improve this further. If you have any ideas, please share them with me. Thanks.

rfaasse replied:
Ahh this rings a bell, that is indeed unfortunate, Indeed since you use const in the lambdas it is the best we can do given the `DofsVectorType`.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:
```diff
@@ -0,0 +1,116 @@
+{
+    auto result        = Vector(rDofs.size());
+    auto get_dof_value = [Step](const auto p_dof) -> double {
+        // Why should we return 0.0 for water pressures?
```
Also here I'm not sure if we should keep the comment

avdg81 replied:
See my previous replies.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:
```diff
@@ -0,0 +1,116 @@
+Vector ExtractSecondTimeDerivatives(const std::vector<Dof<double>*>& rDofs, int Step)
+{
+    auto result                     = Vector(rDofs.size());
+    auto get_second_time_derivative = [Step](const auto p_dof) -> double {
```
If we extract a function, we could extract the first time derivative `Variable` using the cast, and for the second just call `GetTimeDerivative` on the extracted first time derivative. This saves some duplication, especially since we have to do some extra explanation in the comment I think it would help to have that only in one place. We can still create the lambdas for the transforms

avdg81 replied:
Good suggestion. Done.

rfaasse replied:
Nice!

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_dof_utilities.cpp:
```diff
@@ -0,0 +1,53 @@
+    KRATOS_EXPECT_TRUE(ExtractEquationIdsFrom(dofs).empty())
+}
+
+KRATOS_TEST_CASE_IN_SUITE(ExtractingEquationIdsFromDofsYieldsAssociatedIds, KratosGeoMechanicsFastSuite)
```
We are still missing tests for the other 6 utility functions

avdg81 replied:
I have added new unit tests to cover these new utility functions.

rfaasse reviewed (changes requested):
Very impressive @avdg81, this cleans up so many duplications (see also the net negative line count 😄 ) and gives nice examples of how to use the stl library: 
![image](https://github.com/KratosMultiphysics/Kratos/assets/56549273/58777092-dca2-477f-b150-2929ef36636d)

I have some suggestions for improvements and minor clean-up. 

In my opinion. the only one thing that's still missing before we can merge are unit tests for some of the new utility functions.

avdg81 committed:
Processed most of the review feedback

- Moved two function implementations from header file to implementation
  file.
- Put the DOF utility functions in a dedicated `namespace`.
- Replaced a 'raw' loop by an STL algorithm.
- Extracted a utility function that returns the degrees of freedom for a
  collection of nodes and a DOF variable.
- Extracted a local helper function that returns the first time
  derivative of the given degree of freedom.
- Renamed a function parameter.
- Moved a few function bodies around to match the order of the function
  declarations in the header file.

What is still left to be done is to add unit tests for the new utility
functions.

rfaasse reviewed (commented):
Trying to add a new review to remove blocking the PR due to 'Request Changes', but also not approving it.

avdg81 committed:
Added a first unit test for `ExtractDofsFromNodes`

It verifies whether an empty vector is returned when an empty collection
of nodes is supplied.

avdg81 committed:
Added a second unit test for `ExtractDofsFromNodes`

It verifies whether an exception is thrown when attempting to extract a
non-existing DOF from the nodes of an element.

avdg81 committed:
Added a third unit test for `ExtractDofsFromNodes`

Given a 3-noded element, with "DISPLACEMENT_X" as the only degree of
freedom of each node, after extracting the DOFs from the nodes, verify
that:
- The number of returned DOFs equals the number of nodes.
- Each returned DOF pointer is not null.
- Each returned DOF corresponds to "DISPLACEMENT_X".
- Each returned DOF corresponds to the node it is attached to.

avdg81 committed:
Added a unit test for `ExtractSolutionStepValues`

It checks whether the nodal values at the given buffer index are
retrieved as expected using a vector of DOF pointers.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.cpp:
```diff
@@ -0,0 +1,120 @@
+    // instance. However, that class does not give us access to the first time derivative.
+    // Therefore, we need to downcast to a `Variable<double>` first. In practice, we expect this
+    // cast to always succeed, but we will check it here, just in case.
+    auto p_variable = dynamic_cast<const Variable<double>*>(&pDof->GetVariable());
```
For readability, it might be nice to add parentheses here (at first I was wondering why we need the adres of pDof, but it's the pointer to the variable).
```suggestion
    auto p_variable = dynamic_cast<const Variable<double>*>(&(pDof->GetVariable()));
```

avdg81 replied:
Yes, I agree that parentheses make the intention clearer. I have added them.

rfaasse reviewed (commented):
Thank you for processing the review comments! The quality improvement with respect to the original code (i.e. current master branch) is really impressive! 

I have a single small nitpick about parentheses we can add for readability, but it's really minor and apart from the unit tests this branch is ready to be merged from my point of view.

avdg81 committed:
Refactored the unit test for `ExtractSolutionStepValues`

Added a few local helper functions to simplify the test code.

Also modified the API for extracting the DOFs for the given collection
of nodes. Now there are two function templates. The first one receives a
reference-to-const to a range of nodes. The second one receives a pair
of iterators for passing the range of nodes.

avdg81 committed:
Added a unit test for `ExtractFirstTimeDerivatives`

Also refactored the test code.

avdg81 committed:
Renamed three function parameters and one template type name

avdg81 committed:
Added a unit test for `ExtractSecondTimeDerivatives`

avdg81 committed:
Added a first unit test for `ExtractSolutionStepValuesOfUPwDofs`

It checks whether the nodal values are returned when extracting the
values of a displacement component.

Also renamed one of the other unit tests.

avdg81 committed:
Added a second unit test for `ExtractSolutionStepValuesOfUPwDofs`

It checks whether zeroes are returned when extracting the values of the
water pressure, even though the actual values are non-zeroes.

avdg81 committed:
Added a first unit test for `ExtractFirstTimeDerivativesOfUPwDofs`

It checks whether the nodal first derivatives are returned when
extracting the first time derivatives of the displacement DOFs.

avdg81 committed:
Added a second unit test for `ExtractFirstTimeDerivativesOfUPwDofs`

It checks whether zeroes are returned when extracting the first
derivatives of the water pressure, even though the actual derivatives
are non-zeroes.

avdg81 committed:
Added a first unit test for `ExtractSecondTimeDerivativesOfUPwDofs`

It checks whether the nodal second derivatives are returned when
extracting the second time derivatives of the displacement DOFs.

avdg81 committed:
Added a second unit test for `ExtractSecondTimeDerivativesOfUPwDofs`

It checks whether zeroes are returned when extracting the second
derivatives of the water pressure, even though no corresponding variable
has been defined.

avdg81 committed:
Added a pair of parentheses to make the code easier to understand

avdg81 committed:
Merge branch 'master' into geo/12073-refactor-dof-processing

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/dof_utilities.h:
```diff
@@ -23,7 +24,22 @@ namespace Kratos::Geo::DofUtilities
 
-std::vector<Dof<double>*> ExtractDofsFromNodes(const Geometry<Node>& rNodes, const Variable<double>& rDofVariable);
+template <typename InputIt>
+std::vector<Dof<double>*> ExtractDofsFromNodes(InputIt                 NodePtrRangeBegin,
+                                               InputIt                 NodePtrRangeEnd,
+                                               const Variable<double>& rDofVariable)
+{
+    auto result = std::vector<Dof<double>*>{};
+    std::transform(NodePtrRangeBegin, NodePtrRangeEnd, std::back_inserter(result),
+                   [&rDofVariable](const auto& r_node) { return r_node.pGetDof(rDofVariable); });
+    return result;
+}
+
+template <typename NodePtrRange>
+std::vector<Dof<double>*> ExtractDofsFromNodes(const NodePtrRange& rNodePtrs, const Variable<double>& rDofVariable)
+{
+    return ExtractDofsFromNodes(std::begin(rNodePtrs), std::end(rNodePtrs), rDofVariable);
+}
```
Very nice generalization!

rfaasse reviewed (approved):
Thank you for your hard work on the tests, they are very easy to read and cover the extracted functionality nicely! From my point of view this PR is ready to be merged!

WPK4FEM reviewed (approved):
Have read through all. Understood most and am happy with removal of multiple definitions D.O.F. and equationId ordering.
